torch/csrc/jit/passes/alias_analysis.cpp

```diff
@@ -0,0 +1,383 @@
+    const auto hasMutableOutputs = std::any_of(
+        node->outputs().cbegin(),
+        node->outputs().cend(),
+        [this](const Value* output) { return isMutableType(output); });
+
+    // We don't have alias info for this node. Either schematize it, or
+    // add it an analyze* method for it.
+    JIT_ASSERT(!hasMutableOutputs);
+  }
+
+  // Bind formal alias annotation to actual alias sets
+  std::unordered_map<Symbol, AliasInfo> formalToActual;
+  for (size_t i = 0; i < schema.arguments().size(); i++) {
+    const auto& formal = schema.arguments()[i].alias_info();
+    const auto& actualValue = node->inputs().at(i);
+    // Skip if there's no alias annotation
+    if (!formal) {
+      continue;
+    }
+
+    // We don't support composite types for alias analysis yet.
+    JIT_ASSERT(formal->containedTypes().size() == 0);
+    // TODO neither unions nor wildcards make sense on an input. We should
+    // disallow them in function schema
+    JIT_ASSERT(!formal->isWildcard())
+    const auto& formalAlias = formal->set();
+
+    // skip if we've already bound this alias
+    if (formalToActual.count(formalAlias) != 0) {
+      continue;
+    }
+
+    const auto& actualAlias = valueToAlias_.at(actualValue);
+
+    // Bind the formal to the actual
+    formalToActual[formalAlias] = actualAlias;
+
+    // Record all writes
+    for (const auto& alias : actualAlias.sets()) {
+      if (formal->isWrite()) {
+        aliasToWrites_[alias].insert(node);
+      }
+    }
+  }
+
+  // Use the formal-actual mapping to give aliases to the outputs
+  for (size_t i = 0; i < schema.returns().size(); i++) {
+    const auto actual = node->outputs().at(i);
+    if (!isMutableType(actual)) {
+      continue;
+    }
+
+    const auto& formal = schema.returns()[i].alias_info();
+    if (!formal) {
+      // This is a fresh tensor
+      giveAlias(actual, getFreshAlias());
+      continue;
+    }
+
+    // We don't support composite types for alias analysis yet.
+    JIT_ASSERT(formal->containedTypes().size() == 0);
+
+    if (formal->isWildcard()) {
+      giveAlias(actual, AliasInfo::createWildcard());
+      continue;
+    }
+
+    const auto& formalAlias = formal->set();
+    auto outputAlias = formalToActual.at(formalAlias);
+
+    giveAlias(actual, outputAlias);
+  }
+}
+
+void AliasDb::analyzeIf(Node* node) {
+  // For if statements, the alias set of an output is the union of the
+  // alias sets generated by the if and else block
+  const auto trueBlock = node->blocks().at(0);
+  const auto falseBlock = node->blocks().at(1);
+  analyze(trueBlock);
+  analyze(falseBlock);
+
+  for (size_t i = 0; i < node->outputs().size(); i++) {
+    const auto nodeOutput = node->outputs()[i];
+    if (!isMutableType(nodeOutput)) {
+      continue;
+    }
+
+    const auto trueOutput = trueBlock->outputs().at(i);
+    const auto falseOutput = falseBlock->outputs().at(i);
+
+    // If a value is only used in one of the branches, then other branch
```

Comment:
I'm referring to a case that's like:
```
%foo : Tensor = prim::If():
  block0():
    ... do something with foo
    -> (%foo.1)
  block1()
    ... empty
    -> (%foo.2)
```
In this case, the alias analysis pass will never give an alias set to `%foo.2` because it wasn't the output of any node.

Reply:
hmmm, looks like you're right. For some reason I thought the `if` blocks generated a fresh output…